src/vs/workbench/api/browser/mainThreadManagedSockets.ts

```diff
@@ -0,0 +1,178 @@
+		const that = this;
+		const scoketFactory = new class implements ISocketFactory<RemoteConnectionType.Managed> {
+
+			supports(connectTo: ManagedRemoteConnection): boolean {
+				return (connectTo.id === socketFactoryId);
+			}
+
+			connect(connectTo: ManagedRemoteConnection, path: string, query: string, debugLabel: string): Promise<ISocket> {
+				return new Promise<ISocket>((resolve, reject) => {
+					if (connectTo.id !== socketFactoryId) {
+						return reject(new Error('Invalid connectTo'));
+					}
+
+					const factoryId = connectTo.id;
+					that._proxy.$openRemoteSocket(factoryId).then(socketId => {
+						const half: RemoteSocketHalf = {
+							onClose: new Emitter(),
+							onData: new Emitter(),
+							onEnd: new Emitter(),
+						};
+						that._remoteSockets.set(socketId, half);
+
+						ManagedSocket.connect(socketId, that._proxy, path, query, debugLabel, half)
+							.then(
+								socket => {
+									socket.onDidDispose(() => that._remoteSockets.delete(socketId));
+									resolve(socket);
+								},
+								err => {
+									that._remoteSockets.delete(socketId);
+									reject(err);
+								});
+					}).catch(reject);
+				});
+			}
+		};
+		this._registrations.set(socketFactoryId, this._remoteSocketFactoryService.register(RemoteConnectionType.Managed, scoketFactory));
+
+	}
+
+	async $unregisterSocketFactory(socketFactoryId: number): Promise<void> {
+		this._registrations.get(socketFactoryId)?.dispose();
+	}
+
+	$onDidManagedSocketHaveData(socketId: number, data: VSBuffer): void {
+		this._remoteSockets.get(socketId)?.onData.fire(data);
+	}
+
+	$onDidManagedSocketClose(socketId: number, error: string | undefined): void {
+		this._remoteSockets.get(socketId)?.onClose.fire({
+			type: SocketCloseEventType.NodeSocketCloseEvent,
+			error: error ? new Error(error) : undefined,
+			hadError: !!error
+		});
+		this._remoteSockets.delete(socketId);
+	}
+
+	$onDidManagedSocketEnd(socketId: number): void {
+		this._remoteSockets.get(socketId)?.onEnd.fire();
+		this._remoteSockets.delete(socketId);
+	}
+}
+
+interface RemoteSocketHalf {
+	onData: Emitter<VSBuffer>;
+	onClose: Emitter<SocketCloseEvent>;
+	onEnd: Emitter<void>;
+}
+
+export class ManagedSocket extends Disposable implements ISocket {
+	public static connect(
+		socketId: number,
+		proxy: ExtHostManagedSocketsShape,
+		path: string, query: string, debugLabel: string,
+
+		half: {
+			onClose: Emitter<SocketCloseEvent>;
+			onData: Emitter<VSBuffer>;
+			onEnd: Emitter<void>;
+		}
+	): Promise<ManagedSocket> {
+		const socket = new ManagedSocket(socketId, proxy, debugLabel, half.onClose, half.onData, half.onEnd);
+
+		socket.write(VSBuffer.fromString(makeRawSocketHeaders(path, query, debugLabel)));
+
+		const d = new DisposableStore();
+		return new Promise<ManagedSocket>((resolve, reject) => {
+			d.add(socket.onData(d => {
+				if (d.indexOf(socketRawEndHeaderSequence) !== -1) {
+					resolve(socket);
+				}
+			}));
```

Comment:
I think this has a bug if the packets are split in such a way that the ending `\r\n\r\n` sequence is delivered in two packets. Also, I think any data after `\r\n\r\n` needs to be pumped back otherwise it would be dropped.